kani-driver/src/concrete_playback.rs

```diff
@@ -256,156 +281,127 @@ struct UnitTest {
 ///         ..., ] }
 ///     ..., ] }
 /// ```
-/// TODO: this parser should be overhauled once the new Rust CBMC output parser is merged in.
-/// Link to the new Rust parser PR: <https://github.com/model-checking/kani/pull/1433>.
-/// Link to the issue: <https://github.com/model-checking/kani/issues/1477>.
-mod parser {
+mod concrete_vals_extractor {
+    use crate::cbmc_output_parser::{CheckStatus, ParserItem, Property, TraceItem};
     use anyhow::{ensure, Context, Result};
-    use serde_json::Value;
-    use std::fs::File;
-    use std::io::BufReader;
-    use std::path::Path;
 
     pub struct ConcreteVal {
         pub byte_arr: Vec<u8>,
         pub interp_val: String,
     }
 
-    /// Extract concrete values from a failing harness.
-    pub fn extract_concrete_vals(output_filename: &Path) -> Result<Vec<ConcreteVal>> {
-        let cbmc_out = read_cbmc_out(output_filename).with_context(|| {
-            format!("Invalid CBMC output trace file: {}", output_filename.display())
-        })?;
-        let concrete_vals = parse_cbmc_out(&cbmc_out)
-            .context("Failed to parse the CBMC output trace JSON to get concrete values")?;
-        Ok(concrete_vals)
-    }
-
-    /// Read in the CBMC results file and deserialize it to a JSON object.
-    fn read_cbmc_out(results_filename: &Path) -> Result<Value> {
-        let results_file =
-            File::open(results_filename).context("Failed to open CBMC output trace file")?;
-        let reader = BufReader::new(results_file);
-        let json_val = serde_json::from_reader(reader)
-            .context("Could not convert CBMC output trace file into valid JSON")?;
-        Ok(json_val)
-    }
-
-    /// The first-level CBMC output parser. This extracts the result message.
-    fn parse_cbmc_out(cbmc_out: &Value) -> Result<Vec<ConcreteVal>> {
+    /// The first-level extractor. Traverses processed items to find properties.
+    pub fn extract_from_processed_items(
+        processed_items: &[ParserItem],
+    ) -> Result<Vec<ConcreteVal>> {
         let mut concrete_vals: Vec<ConcreteVal> = Vec::new();
-        let cbmc_out_arr = cbmc_out
-            .as_array()
-            .with_context(|| format!("Expected this CBMC output to be an array: {}", cbmc_out))?;
-        let mut have_parsed_assert_fail = false;
-        for general_msg in cbmc_out_arr {
-            let result_msg = &general_msg["result"];
-            if !result_msg.is_null() {
-                let result_arr = result_msg.as_array().with_context(|| {
-                    format!("Expected this CBMC result object to be an array: {}", result_msg)
-                })?;
-                for result_val in result_arr {
-                    parse_result(result_val, &mut concrete_vals, &mut have_parsed_assert_fail)?;
+        let mut extracted_assert_fail = false;
+        for processed_item in processed_items {
+            if let ParserItem::Result { result } = processed_item {
```

Comment:
You could use a helper function function here to retrieve the result first and avoid a nested loop. If there is no result in `processed_items`, that's an error.

Reply:
Good point! Done